src/qt/clientmodel.cpp

```diff
@@ -116,9 +116,22 @@ int ClientModel::getNumBlocks() const
 
 uint256 ClientModel::getBestBlockHash()
 {
+    // Lock order must be: first `cs_main`, then `m_cached_tip_mutex`.
+
+    uint256 tip{WITH_LOCK(m_cached_tip_mutex, return m_cached_tip_blocks)};
+
+    if (!tip.IsNull()) {
+        return tip;
+    }
+
+    tip = m_node.getBestBlockHash(); // Will lock `cs_main` (and release it).
+
     LOCK(m_cached_tip_mutex);
+    // We checked that `m_cached_tip_blocks` is not null above, but then we
+    // released the mutex `m_cached_tip_mutex`, so it could have changed in the
+    // meantime. Thus, check again.
     if (m_cached_tip_blocks.IsNull()) {
-        m_cached_tip_blocks = m_node.getBestBlockHash();
+        m_cached_tip_blocks = tip;
     }
     return m_cached_tip_blocks;
```

Comment:
If the requirement to check `IsNull` twice is dropped, then I think this whole function can be written shorter in just 4 lines of code:

```cpp
     uint256 tip{WITH_LOCK(m_cached_tip_mutex, return m_cached_tip_blocks)};
     if (!tip.IsNull()) return tip;

     tip = m_node.getBestBlockHash(); // Will lock `cs_main` (and release it). This is needed to be done early to not violate the implicit lock order cs_main -> m_cached_tip_mutex

     return WITH_LOCK(m_cached_tip_mutex, return m_cached_tip_blocks = tip);

Reply:
The above snippet can end up setting/overwriting `m_cached_tip_blocks` if it is not null. I.e. if it is null during the `if`, later some other thread sets it to some non null value (while we are calling `m_node.getBestBlockHash()`) and we end up overwriting what the other thread set.

Maybe that is ok in this case or maybe it cannot happen due to some constraints outside of this function. This code is new to me so I played safe :)

Btw `if (A) return B;` has two disadvantages compared to putting the `return` on a separate line:
1. it is not possible to set a breakpoint on `return B;` only. For example if the condition is evaluated 1000 times, but `A` is `true` only a few times and one wants to break when `return B;` is executed.
2. it will skew line-based code coverage.